apps/rewards/src/invite/rewardSender.service.ts

```diff
@@ -0,0 +1,156 @@
+import { InviteRewardRepository } from './inviteReward.repository'
+
+const WEI_PER_UNIT = 1000000000000000000
+
+interface WatchedInvite {
+  inviteId: string
+  txHash: string
+  sentAt: number
+  markAsDeadLetterIfFailed: boolean
+}
+
+@Injectable()
+export class RewardSenderService {
+  watchedInvites: Set<WatchedInvite>
+
+  constructor(
+    private readonly inviteRewardRepository: InviteRewardRepository,
+    private readonly relayerProxyService: RelayerProxyService,
+    @Inject(appConfig.KEY)
+    private readonly appCfg: AppConfig,
+    private readonly contractKit: ContractKit,
+    private readonly logger: KomenciLoggerService
+  ) {
+    this.watchedInvites = new Set()
+  }
+
+  async wasTxSentSuccesfully(txHash: string) {
+    const tx = await this.contractKit.web3.eth.getTransaction(txHash)
+    return Boolean(tx.blockHash)
+  }
+
+  @Cron(CronExpression.EVERY_SECOND)
+  async checkWatchedInvites() {
+    await Promise.all(
+      [...this.watchedInvites].map(async invite => {
+        if (await this.wasTxSentSuccesfully(invite.txHash)) {
+          await this.inviteRewardRepository.update(invite.inviteId, {
+            state: RewardStatus.Completed
+          })
+          this.watchedInvites.delete(invite)
+        } else if (
+          Date.now() - invite.sentAt >
+          this.appCfg.transactionTimeoutMs
+        ) {
+          await this.inviteRewardRepository.update(invite.inviteId, {
+            state: invite.markAsDeadLetterIfFailed
+              ? RewardStatus.DeadLettered
+              : RewardStatus.Failed
+          })
+          this.watchedInvites.delete(invite)
+        }
+      })
+    )
+  }
+
+  @Cron(CronExpression.EVERY_MINUTE)
+  async checkFailedInvites() {
+    await this.inviteRewardRepository.manager.transaction(
+      async (entityManager: EntityManager) => {
+        const repository = entityManager.getRepository<InviteReward>(
+          'invite_reward'
+        )
+        const invites = await repository
+          .createQueryBuilder()
+          .setLock('pessimistic_read')
+          .where({
+            state: In([RewardStatus.Submitted, RewardStatus.Failed]),
+            createdAt: Raw(
+              alias =>
+                `${alias} <= current_timestamp - (30 ||' minutes')::interval`
+            )
+          })
+          .getMany()
+        if (invites.length > 0) {
+          this.logger.log(`Found failed or submitted invites ${invites.length}`)
+        }
+        await Promise.all(
+          invites.map(async invite => {
+            if (
+              invite.state === RewardStatus.Submitted &&
+              (await this.wasTxSentSuccesfully(invite.rewardTxHash))
+            ) {
+              await repository.update(invite.id, {
+                state: RewardStatus.Completed
+              })
+            } else {
+              await this.sendInviteReward(invite, true, repository)
+            }
+          })
+        )
+      }
+    )
+  }
+
+  async getRewardInCelo() {
```

Comment:
It may be more efficient to calculate this once for a given batch of transactions instead of per each one (given the price shouldn't change too much)